app/src/main/kotlin/at/bitfire/davdroid/ui/AppSettingsScreen.kt

```diff
@@ -492,15 +506,118 @@ fun AppSettings_Integration(
     )
 
     val context = LocalContext.current
+    val pm = context.applicationContext.packageManager
+
+    var distributor by remember { mutableStateOf(UnifiedPush.getSavedDistributor(context)) }
+    val distributors = remember { UnifiedPush.getDistributors(context) }
+    LaunchedEffect(Unit) {
+        UnifiedPush.getAckDistributor(context)?.let {
+            UnifiedPush.registerApp(context)
+        }
+    }
+
+    var showingDistributorDialog by remember { mutableStateOf(false) }
+    if (showingDistributorDialog) {
+        var selectedDistributor by remember { mutableStateOf(distributor) }
+
+        AlertDialog(
+            onDismissRequest = { showingDistributorDialog = false },
+            confirmButton = {
+                TextButton(
+                    enabled = selectedDistributor != null,
+                    onClick = {
+                        UnifiedPush.saveDistributor(context, selectedDistributor!!)
+                        UnifiedPush.registerApp(context)
+                        distributor = selectedDistributor
+                        showingDistributorDialog = false
+                    }
+                ) { Text(stringResource(android.R.string.ok)) }
+            },
+            dismissButton = {
+                TextButton(
+                    onClick = { showingDistributorDialog = false }
+                ) { Text(stringResource(android.R.string.cancel)) }
+            },
+            title = {
+                Text(stringResource(R.string.app_settings_unifiedpush_endpoint_choose))
+            },
+            text = {
+                LazyColumn(modifier = Modifier.fillMaxWidth()) {
+                    if (distributors.isEmpty()) item {
+                        Text(stringResource(R.string.app_settings_unifiedpush_endpoint_none))
+                    }
+
+                    items(distributors) { distributor ->
+                        val applicationInfo = remember(distributor) {
+                            try {
+                                pm.getApplicationInfo(distributor, 0)
+                            } catch (_: PackageManager.NameNotFoundException) {
+                                null
+                            }
+                        }
+                        val applicationName = remember(applicationInfo) {
+                            applicationInfo?.let { pm.getApplicationLabel(applicationInfo).toString() }
+                        }
+                        val applicationIcon = remember(applicationInfo) {
+                            applicationInfo?.let { pm.getApplicationIcon(applicationInfo) }
+                        }
```

Comment:
If really necessary, we can call this here, but I think everything that generates data (like the app name) should be done in the model.